### PR TITLE
Setting outgoing protocol

### PR DESCRIPTION
**Pull request by IdavalapatiRamanjaneyulu**

### CODE REVIEW

File: kwikapi/api.py
Comment (jaswanth098) [inline kwikapi/api.py:523]: @RamanjaneyuluIdavalapati What is this protocol used for??

Comment (IdavalapatiRamanjaneyulu) [reply kwikapi/api.py:523]: @jaswanth098 We use that protocol to serialise the response/result.
By default JSON protocol will be used to serialise and deserialise the incoming(client sends data to the server) and outgoing(server sends data to the client) data.

We can change this default protocol. [Ref](https://github.com/deep-compute/kwikapi/blob/master/README.md#kwikapi-also-supports-defining-specific-protocol-for-specific-request)

But this protocol will be used to serialise and deserialise for both incoming and outgoing data.
I have supported here such that the user can change the protocol for outgoing data.

Comment (jaswanth098) [reply kwikapi/api.py:523]: got it thanks for clarification

**Review by jaswanth098 — APPROVED** [review]

**Comment by prashanthellina** [comment]
@RamanjaneyuluIdavalapati Please undo this merge. We should not be using request.headers as a mechanism to signal to kwikapi code that the response protocol is raw type. For this, we can use request.response.headers (which is not there today). Please add `headers` as an abstract property to `BaseResponse` and implement support in `TornadoResponse` and `DjangoResponse`.

If the above is done, instead of `_find_request_protocol`, we can have `_find_response_protocol` which will examine `request.response` instead of `request`.

In addition, we can take all the headers mentioned in request.response.headers and set them in the raw response so they are sent back to the caller.

Please call me if this isn't clear.

**Comment by IdavalapatiRamanjaneyulu** [comment]
@prashanthellina How do I undo/revert the merge and commit here? I have seen this https://stackoverflow.com/questions/2389361/undo-a-git-merge-that-hasnt-been-pushed-yet. This will work in local.

After doing the above should I commit again?

**Comment by prashanthellina** [comment]
@RamanjaneyuluIdavalapati No idea. Haven't done that before.

**Comment by jaswanth098** [comment]
@RamanjaneyuluIdavalapati you can see a revert option above beside the merge that amani has done

**Comment by IdavalapatiRamanjaneyulu** [comment]
@jaswanth098 I tried it, but it is creating another commit.

**Comment by jaswanth098** [comment]
Yes that is the revert commit that has to be applied in github, once a commit is made to the github it cannot be removed you can only make one more commit by removing all those changes that is what the above revert is about

**Comment by IdavalapatiRamanjaneyulu** [comment]
@jaswanth098 @prashanthellina If that is the case (commit can't be deleted), let me implement response headers and give pull request instead of reverting.

It is because there is of no use with revert other than making duplicate commits.

**Comment by IdavalapatiRamanjaneyulu** [comment]
@jaswanth098 @prashanthellina This is working https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github to remove the commit. But it will remove only visibility of the commit from the Github. If you have the commit id then you can still access the commit.

**Comment by prashanthellina** [comment]
@RamanjaneyuluIdavalapati That works.

**Comment by IdavalapatiRamanjaneyulu** [comment]
@prashanthellina will work on this tomorrow. Do I have write access to this repo? If I have I can simply remove the commit. Not sure will this work(removing) for pull request also?

**Comment by prashanthellina** [comment]
@RamanjaneyuluIdavalapati You have write access and that allows you to edit the code in the repo.

**Comment by IdavalapatiRamanjaneyulu** [comment]
@prashanthellina I don't have permission to push directly. May be this is because the branch is protected.

Screenshot 2019-05-14 at 11 58 59 AM

**Comment by jaswanth098** [comment]
@RamanjaneyuluIdavalapati No one has permission to push to master you should start a pull request to master branch from a different branch

**Comment by IdavalapatiRamanjaneyulu** [comment]
@jaswanth098 Is this because of Github or we set up like this? If we set up like this can't we have permission for this push alone (Change the set up until this pull gets merge)?

If we can't able to change it then I will give a pull request from different branch. Not sure whether we can able to merge or not. Not sure will this delete the commits.

**Comment by jaswanth098** [comment]
@RamanjaneyuluIdavalapati This is something that we did intentionally because no one pushes some buggy code to master with a review
Right now I don't have permission to change this setting.
anyway this is an enhancement that you are gonna do right why don't you do it on top of it instead of doing a revert in the code

**Comment by IdavalapatiRamanjaneyulu** [comment]
@jaswanth098 I have checked on my personal Github repository by pushing to different branch. When I tried to create a pull request it is saying no changes(even there are changes between commits) and not accepting the pull request.

> why don't you do it on top of it instead of doing a revert in the code

You mean keep the commit as it is and proceed to implement further features?

**Comment by jaswanth098** [comment]
> @jaswanth098 I have checked on my personal Github repository by pushing to different branch. When I tried to create a pull request it is saying no changes(even there are changes between commits) and not accepting the pull request.
> 
> > why don't you do it on top of it instead of doing a revert in the code
> 
> You mean keep the commit as it is and proceed to implement further features?

yes

**Comment by IdavalapatiRamanjaneyulu** [comment]
> anyway this is an enhancement that you are gonna do right 

@jaswanth098 I have checked this on my personal Github repo. It worked there when I directly push to master. That is the reason I told this is working.

**Comment by IdavalapatiRamanjaneyulu** [comment]
> @jaswanth098 @prashanthellina This is working https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github to remove the commit. But it will remove only visibility of the commit from the Github. If you have the commit id then you can still access the commit.

@jaswanth098 @prashanthellina Example for the above.
I have given 3 commits for the repo https://github.com/chowdaryidavalapati/removecommit.
But you can only find 2 because I have removed one. This will be removed from the history but still can be accessed if you have the commit id.
**3rd commit:** https://github.com/chowdaryidavalapati/removecommit/tree/bcafd782b338817cc9285fc7618d2f8b35922dfa

But anyway will not worry about this and will continue to implement further features as you suggested.
I have provide this because this may help in the future.

**Pull request merged**
